c2rust-analyze/src/rewrite/ty.rs

```diff
@@ -0,0 +1,370 @@
+use rustc_middle::ty::subst::GenericArg;
+use rustc_middle::ty::{self, ReErased, Ty, TyCtxt, TyKind};
+use rustc_span::Span;
+
+/// A label for use with `LabeledTy` to indicate what rewrites to apply at each position in a type.
+#[derive(Clone, Copy, PartialEq, Eq, Debug, Hash)]
+struct RewriteLabel {
+    /// Rewrite a raw pointer, whose ownership and quantity have been inferred as indicated.
+    ty_desc: Option<(Ownership, Quantity)>,
+    /// If set, a child or other descendant of this type requires rewriting.
+    descendant_has_rewrite: bool,
+}
+
+type RwLTy<'tcx> = LabeledTy<'tcx, RewriteLabel>;
+
+/// Given an `LTy`, which is labeled with `PointerId`s, determine which rewrites to apply based on
+/// the permissions and flags inferred for each `PointerId`.
+fn relabel_rewrites<'tcx>(
+    asn: &Assignment,
+    lcx: LabeledTyCtxt<'tcx, RewriteLabel>,
+    lty: LTy<'tcx>,
+) -> RwLTy<'tcx> {
+    lcx.relabel_with_args(lty, &mut |lty, args| {
+        let ty_desc = if lty.label.is_none() {
+            None
+        } else {
+            let perms = asn.perms()[lty.label];
+            let flags = asn.flags()[lty.label];
+            // TODO: if the `Ownership` and `Quantity` exactly match `lty.ty`, then `ty_desc` can
+            // be `None` (no rewriting required).  This might let us avoid inlining a type alias
+            // for some pointers where no actual improvement was possible.
+            Some(type_desc::perms_to_desc(perms, flags))
+        };
+        let descendant_has_rewrite = args.iter().any(|child| {
+            let has_rewrite = child.label.ty_desc.is_some();
+            has_rewrite || child.label.descendant_has_rewrite
+        });
+        RewriteLabel {
+            ty_desc,
+            descendant_has_rewrite,
+        }
+    })
+}
+
+/// Extract arguments from `hir_ty` if it corresponds to the tcx type `ty`.  If the two types
+/// differ in shape (for example, if `hir_ty` is `_` or a type alias), then the result is `None`.
+/// If they have the same shape, the result is a list of `hir::Ty`s of arguments, where "argument"
+/// is defined the same way as for the `args` field of `LabeledTy`.
+fn deconstruct_hir_ty<'a, 'tcx>(
+    ty: Ty<'tcx>,
+    hir_ty: &'a hir::Ty<'tcx>,
+) -> Option<Vec<&'a hir::Ty<'tcx>>> {
+    /// Match a specific primitive type against a `hir::Ty`.  Returns `Some([])` on success (since
+    /// primitive types have no arguments) or `None` on failure.
+    fn do_prim<'a, 'tcx>(
+        expect_prim: hir::PrimTy,
+        hir_ty: &hir::Ty<'tcx>,
+    ) -> Option<Vec<&'a hir::Ty<'tcx>>> {
+        let path = match hir_ty.kind {
+            hir::TyKind::Path(hir::QPath::Resolved(_, path)) => path,
+            _ => return None,
+        };
+        let prim = match path.res {
+            Res::PrimTy(prim) => prim,
+            _ => return None,
+        };
+        if prim != expect_prim {
+            return None;
+        }
+        Some(Vec::new())
+    }
```

Comment:
Why not return `bool` or `Option<()>`?

Reply:
`do_prim` is used to factor out a bunch of case bodies in the match below, and the result type of that match is `Option<Vec<...>>` (because the non-primitive cases return non-empty lists when the result is `Some`).